src/handle_commit.cxx

```diff
@@ -300,14 +300,27 @@ void raft_server::commit_app_log(ulong idx_to_commit,
     std::list< ptr<commit_ret_elem> > async_elems;
     if (need_to_handle_commit_elem) {
         std::unique_lock<std::mutex> cre_lock(commit_ret_elems_lock_);
-        bool match_found = false;
+        /// Sometimes user can batch requests to RAFT: for example send 30
+        /// append entries requests in a single batch. For such request batch
+        /// user will receive a single response: all was successful or all
+        /// failed. Obviously we don't need to add info about responses
+        /// (commit_ret_elems) for 29 requests from batch and need to do it only
+        /// for 30-th request. precommit_index is exact value which identify ID
+        /// of the last request from the latest batch. So if we commiting this
+        /// last request and for some reason it was not added into
+        /// commit_ret_elems in the handle_cli_req method (logical race
+        /// condition) we have to add it here. Otherwise we don't need to add
+        /// anything into commit_ret_elems_, because nobody will wait for the
+        /// responses of the intermediate requests from requests batch.
+        bool need_to_check_commit_ret = sm_idx == pc_idx;
```

Comment:
Thanks for this fix. To make sure that I understand this condition correctly, can you confirm the below?

Let's say the current committed index is 10 (for all pre-commit, quick-commit, and sm-commit), and two threads (T1 and T2) are calling `handle_cli_req` with exactly one log. What will happen is:
1. T1 calls `handle_cli_req` first, `precommit_index_` becomes 11.
2. Before T1 creates the corresponding `commit_ret_elem`, the commit thread wakes up earlier.
3. Since `sm_idx == pc_idx` in this case, the commit thread will create it.
4. However, T2 cannot interfere before step (3), because of the lock held by `handle_cli_req_prelock`. T2 can append its log only after the `commit_ret_elem` for 11 is created.

So, if a thread order inversion happens, only one user thread is involved at a time. When the commit thread wakes up, there is no case that there are multiple `handle_cli_req` calls whose `commit_ret_elem`s are not created yet. Is that right?

Reply:
Actually I didn't think about such case, but yes `handle_cli_req` guarded by mutex (or recursive mutex for some reason...) in `handle_cli_req_prelock`:  https://github.com/ClickHouse-Extras/NuRaft/blob/1707a7572aa66ec5d0a2dbe2bf5effa3352e6b2d/src/handle_client_request.cxx#L40-L53.

Reply:
BTW I cannot catch "thread inversion" even under heavy load. But according to the code it should be possible.

Reply:
Thanks, let me merge it.